### PR TITLE
feat: add export container

### PR DESCRIPTION
**Pull request by lstocchi**
### What does this PR do?

It adds the export container action

### Screenshot / video of UI

export_container

### What issues does this PR fix or reference?

it is part of #6355 

### How to test this PR?

1. try to export a container and see it is actually saved in the filesystem

- [x] Tests are covering the bug fix or the new feature

### CODE REVIEW

File: packages/renderer/src/lib/container/ContainerActions.svelte [hunk 1]
```diff
@@ -202,6 +209,12 @@ if (dropdownMenu) {
     menu="{dropdownMenu}"
     detailed="{detailed}"
     icon="{faArrowsRotate}" />
+  <ListItemButtonIcon
+    title="Export Container"
```
Comment (benoitf) [inline packages/renderer/src/lib/container/ContainerActions.svelte:213]: the full length is 'Exports container's filesystem contents as a tar archive and saves it on the local machine'

I don't know if we can add a custom tooltip there ?

File: packages/renderer/src/stores/export-container-store.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+import type { Writable } from 'svelte/store';
+import { writable } from 'svelte/store';
+
+import type { ContainerInfoUI } from '../lib/container/ContainerInfoUI';
+
+/**
+ * Defines the store used to define the container to export.
+ */
+export const exportContainerInfo: Writable<ContainerInfoUI> = writable();
```
Comment (benoitf) [inline packages/renderer/src/stores/export-container-store.ts:27]: I know it's a dummy one but maybe add a test there ?

**Review by benoitf — COMMENTED** [review]
the export on more than one container will come in follow-up PRs ?

File: packages/main/src/plugin/container-registry.ts [hunk 3]
```diff
@@ -2405,23 +2404,10 @@ export class ContainerProviderRegistry {
...
-    const regexFile = new RegExp(`${containerFile}\\s+\\(+\\d+\\)$`);
-    const entries = await fs.promises.readdir(options.outputDirectory, { withFileTypes: true });
-    const similarContainers = entries
-      .filter(entry => entry.isFile())
-      .filter(file => file.name === containerFile || regexFile.test(file.name));
-
-    if (similarContainers.length > 0) {
-      containerFile += ` (${similarContainers.length})`;
-    }
-
```
Comment (benoitf) [inline packages/main/src/plugin/container-registry.ts:2420]: if the file already exists we overwrite the content now ?

Comment (lstocchi) [reply packages/main/src/plugin/container-registry.ts:2420]: Yes.
In my first implementation i was using the openDialog to make the user select the output folder + there was a textbox where to insert the name of the output file (then i was combining the two and if the file already existed i was adding a `(1)` suffix). 
Now i switched to a saveDialog. So if you try to use the name of an already existing file, it already shows you a warning "do you want to replace the existing file?" (or something like that), so the user is already informed. This is why i removed this part.

Comment (benoitf) [reply packages/main/src/plugin/container-registry.ts:2420]: got it

**Review by benoitf — APPROVED** [review]

**Comment by lstocchi** [comment]
> the export on more than one container will come in follow-up PRs ?

I missed this comment.
Yes, i'm not actually sure how this should work. At first i thought about asking the user to select the folder and download any single container file by using its own name (and rename it if it was already in the filesystem - like the code removed above). But now that i used the saveDialog i thought that the user can create a sort of archive file or folder and everything will be downloaded in it with the original container file. Any preference?
BTW i was working on the import action so atleast we have a working export-import on a single file.

**Comment by cdrage** [comment]
May have to follow up with UX team, but I believe if it's non-form, it should be a modal instead similar to push images. 

Anything more complicated than 3-4 actions I believe was suppose to be a form, but anything very simple is a modal. 

We could use the same template for push images for this.

**Comment by lstocchi** [comment]
> May have to follow up with UX team, but I believe if it's non-form, it should be a modal instead similar to push images.
> 
> Anything more complicated than 3-4 actions I believe was suppose to be a form, but anything very simple is a modal.
> 
> We could use the same template for push images for this.

Are you talking about the multiple export or this single one? I didn't use a modal here bc there is the saveDialog which opens another dialog so it looked to me too annoying to have two opened up

**Comment by benoitf** [comment]
> But now that i used the saveDialog i thought that the user can create a sort of archive file or folder and everything will be downloaded in it with the original container file. Any preference?
BTW i was working on the import action so atleast we have a working export-import on a single file.

I thought the command was allowing multiple containers (like podman save where you can provide multiple images)

on a preference, I'll go to something simple like if I select a folder it'll export each selected container to that folder 


and then on the import if we select multiple files it'll import multiple containers

**Comment by lstocchi** [comment]
> I thought the command was allowing multiple containers (like podman save where you can provide multiple images)

No, it looks to me that it only accept one item so we need to call it for every single selected item

> on a preference, I'll go to something simple like if I select a folder it'll export each selected container to that folder
> 
> and then on the import if we select multiple files it'll import multiple containers

Ok

File: packages/renderer/src/lib/container/ContainerExport.svelte [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            <Input
+              class="grow mr-2"
+              name="{container.id}"
+              readonly
+              value="{outputTarget}"
+              id="input-export-container-name"
+              aria-invalid="{invalidFolder}" />
+            <Button
+              on:click="{() => selectFolderPath()}"
+              icon="{faFolderOpen}"
```
Comment (deboer-tim) [inline packages/renderer/src/lib/container/ContainerExport.svelte]: We don't use icons on Browse buttons anywhere else. There's a new design where we wouldn't even have the Button, but in the meantime I think we should be consistent and not include the icon.

File: packages/renderer/src/lib/container/ContainerExport.svelte [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+{#if container}
+  <FormPage title="Export container {container.name}">
+    <svelte:fragment slot="icon">
+      <i class="fas fa-download fa-2x" aria-hidden="true"></i>
+    </svelte:fragment>
+
+    <div slot="content" class="flex flex-col min-w-full h-fit p-5">
+      <div class="bg-charcoal-600 px-6 py-4 space-y-2 lg:px-8 sm:pb-6 xl:pb-8">
+        <div>
```
Comment (deboer-tim) [inline packages/renderer/src/lib/container/ContainerExport.svelte:95]: When I test there is a double-underline: one in the Input component, and one all the way underneath the Input and Button. This second one should be removed.

File: packages/renderer/src/lib/container/ContainerExport.svelte [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  container = $exportContainerInfo;
+  if (!container) {
+    // go back to containers list
+    router.goto('/containers/');
+  }
+});
+
+async function selectFolderPath() {
+  const result = await window.saveDialog({
+    title: 'Select the directory where to export the container content',
```
Comment (deboer-tim) [inline packages/renderer/src/lib/container/ContainerExport.svelte:37]: By default the output folder is 'Unititled'. Could we change this to the name of the image (or at least be unique)?

File: packages/renderer/src/lib/container/ContainerExport.svelte [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+              aria-invalid="{invalidFolder}" />
+            <Button
+              on:click="{() => selectFolderPath()}"
+              icon="{faFolderOpen}"
+              title="Open dialog to select the output file"
+              aria-label="Select output file">Browse ...</Button>
+          </div>
+          <div class="pt-5 border-zinc-600 border-t-2"></div>
+          <Button
+            on:click="{() => exportContainer()}"
```
Comment (deboer-tim) [inline packages/renderer/src/lib/container/ContainerExport.svelte:111]: When I click on the button, nothing happens. I can click on it multiple times, and then suddenly the form disappears. I don't know if anything worked until I go to the tasks view and see multiple export jobs doing the same thing. Button should be inProgress once you click and there should be some indication that a background task has been started before the form disappears.

**Review by deboer-tim — CHANGES_REQUESTED** [review]
Function worked great 👍🏼 but noted a few usability/UI issues.

**Comment by lstocchi** [comment]
> Function worked great 👍🏼 but noted a few usability/UI issues.

I should have fixed everything. Please give it another try!! Thanks!

**Comment by lstocchi** [comment]
@benoitf clicked the wrong button, i wanted to request @deboer-tim review. But you can give it another look if you want 😆

**Review by benoitf — APPROVED** [review]

**Review by deboer-tim — APPROVED** [review]
Thanks for making the changes 👍🏼. Working perfectly for me, and the only usability issue is that the form disappears and user doesn't know if it's done / running in the background / what happened, but that is not export-specific and we already have issues open for that (e.g. #5505).

**Pull request merged**